docs/lazy-adr/adr-004-core-types.md

```diff
@@ -0,0 +1,171 @@
+# ADR 004: Blockchain Core Data Types
+
+## Changelog
+
+- {date}: {changelog}
+
+## Context
+
+> This section contains all the context one needs to understand the current state, and why there is a problem. It should be as succinct as possible and introduce the high level idea behind the solution.
+
+This document describes the core data structures of any Optimint-powered blockchain.
+
+## Alternative Approaches
+
+> This section contains information around alternative options that are considered before making a decision. It should contain a explanation on why the alternative approach(es) were not chosen.
+
+## Decision
+
+We design the core data types as minimalistic as possible, i.e. they only contain the absolute necessary 
+data for an optimistic rollup to function properly. 
+If there are any additional fields that conflict with above's claimed minimalism, then they are necessarily inherited 
+by the ABCI imposed separation between application state machine and consensus/networking (often also referred to as ABCI-server and -client).
+Where such tradeoffs are made, we explicitly comment on them.
+
+## Detailed Design
+
+### Transactions
+
+In Optimint, like in Tendermint, Transactions are just an opaque slice of bytes:
+
+```go
+type Tx []byte
+type Txs []Tx
+```
+
+If necessary `Tx` could be turned into a struct. Currently, there is no need for that though.
+
+### Block Header
+
+```go
+type Header struct {
+    // Block and App version
+    Version Version 
+    // TODO this is redundant; understand if
+    // anything like the ChainID in the Header is 
+    // required for IBC though.
+    NamespaceID [8]byte  
+    
+    Height  uint64               
+    Time    uint64 // time in tai64 format           
+    
+    // prev block info
+    LastHeaderHash [32]byte 
+    
+    // hashes of block data
+    LastCommitHash     [32]byte // commit from aggregator(s) from the last block
+    DataHash           [32]byte // Block.Data root aka Transactions
+    ConsensusHash      [32]byte // consensus params for current block
```

Comment:
Forgot to define the consensus params. They can also be changed via ABCI and hence I propose to keep them the same or at least very close to tendermint's consensus params: https://github.com/tendermint/tendermint/blob/1f01e5d72667e8ce4507fd6a602bb617451f32ab/proto/tendermint/types/params.proto#L13 

Note: We should only include this hash if it doesn't require a different fraud-proof type.

Reply:
So while it is possible to modify the ConsensusParams via ABCI, I do not see that happening in the SDK (only in simapp / during Initchain). So I think we should (for now) treat them as constants. This means we should clearly communicate this app-devs though. Maybe we should even ensure programmatically that the ConsensusParams are not modified in the same version of the state-machine / app inside of optimint.